### PR TITLE
Performance improvement

### PR DESCRIPTION
**Pull request by peteraisher**
Use `UnsafeMutablePointer` instead of `ArraySlice` during `KDTree.init`.

Performance improvement of approximately 4x for tree build.

Performance comparison Array vs UnsafeMutablePointer

### CODE REVIEW

**Comment by Bersaelor** [comment]
Wow, those are impressive results!

We are going down to c-level arrays during the tree creation though, which does add a higher barrier with people to understand the code. 
I started programming in C , so it I'm into it, but give me a day to think about it :)

( @hsnetzer what is your opinion?)

**Comment by peteraisher** [comment]
If readability/understandability is important, I imagine it would be possible to wrap the scary bits of the `UnsafeMutablePointer` code into a custom struct (say `FastArray`) that behaves exactly like `ArraySlice` for the methods that are actually called during tree building.
That ought to retain (most of) the performance advantage of using pointers without sacrificing code readability.

After all, `ArraySlice` is already a little confusing (and un-Swift-y) to the uninitiated, being non-zero-indexed, not owning it’s own memory, etc.

**Comment by chriseidhof** [comment]
To me, it seems like it isn't a huge change in code, and if well documented, it might actually be very useful for people when reading the code. It's not an uncommon thing to resort to UnsafeMutablePointer when you need performance. I also like that it's self-contained (not exposed to the user).

File: Sources/KDTree.swift [hunk 1]
```diff
@@ -18,93 +18,121 @@ public enum KDTree<Element: KDTreePoint> {
...
-        else {
-            var median = values.count / 2
+        let count = values.count
+        
+        let pointer = UnsafeMutablePointer<Element>.allocate(capacity: count)
+        
+        // copy values from the array
+        pointer.initialize(from: values, count: count)
+        
+        self = KDTree(values: pointer, startIndex: 0, endIndex: count)
```
Comment (peteraisher) [inline Sources/KDTree.swift]: oops, this should be
```suggestion
        self = KDTree(values: pointer, startIndex: 0, endIndex: count, depth: depth)
```

File: Sources/KDTree.swift [hunk 2]
```diff
@@ -18,93 +18,121 @@ public enum KDTree<Element: KDTreePoint> {
...
-          
-            let leftTree = KDTree(values: selected[0..<median], depth: depth+1)
-            let rightTree = KDTree(values: selected[median+1..<selected.count], depth: depth+1)
             
-            self = .node(left: leftTree, value: selected[median],
-                         dimension: currentSplittingDimension, right: rightTree)
+            // copy values from the array
+            pointer.initialize(from: start, count: count)
+            
+            return KDTree(values: pointer, startIndex: 0, endIndex: count)
```
Comment (peteraisher) [inline Sources/KDTree.swift]: this should also pass through the depth parameter
```suggestion
            return KDTree(values: pointer, startIndex: 0, endIndex: count, depth: depth)
```

**Comment by hsnetzer** [comment]
I agree that the current use of ArraySlice and inout isn't the Swiftiest thing. UnsafeMutablePointer actually makes the code a bit more concise and consistent. I don't think 'public init(values: ArraySlice<Element>' is needed anymore if we go this way. 

I tested UnsafeMutablePointer against just using inout Array and its still 3x faster. What makes it so fast?

**Comment by Bersaelor** [comment]
Thanks for your replies, I agree. 

@peteraisher : @hsnetzer is correct, we can probably remove `public init(values: ArraySlice<Element>` as part of this PR.

**Comment by Bersaelor** [comment]
> I tested `UnsafeMutablePointer` against just using `inout [Element]` and `UnsafeMutablePointer` is 3x faster. What makes it so fast?

That probably has to do with the way Swift Arrays are implemented. If the element is pretty small (like < 8 Byte, but I don't know the specific number), then the elements will be stored inline similar to a C array.
If the elements are larger they are stored separately, and the only thing stored in the Swift Array is a pointer to the value that is stored somewhere else, in our case our KDTreePoints can get pretty big and therefore the Swift arrays has pretty bad memory access pattern compard to the plain `UnsafeMutablePointer` which is a linear storage of points in the memory.
I remember a WWDC lecture with pretty graphs about this (maybe [this one](https://developer.apple.com/videos/play/wwdc2018/229/) but probably it was a year before that ). 

Anyway, gtm!

**Pull request merged**